abdiff/core/run_ab_transforms.py

```diff
@@ -1 +1,187 @@
 """abdiff.core.run_ab_transforms"""
+
+import logging
+import os
+import re
+import time
+from datetime import timedelta
+from pathlib import Path
+from time import perf_counter
+
+import docker
+from docker.models.containers import Container
+
+from abdiff.config import Config
+from abdiff.core.exceptions import DockerContainerRuntimeExceededTimeoutError
+from abdiff.core.utils import create_subdirectories, update_or_create_run_json
+
+CONFIG = Config()
+
+logger = logging.getLogger(__name__)
+
+
+def run_ab_transforms(
+    run_directory: str,
+    image_name_a: str,
+    image_name_b: str,
+    input_files: list[str],
+    docker_client: docker.client.DockerClient | None = None,
+    timeout: int = 180,
+) -> tuple[list[str], ...]:
+    """Run Docker containers with versioned images of Transmogrifier."""
+    start_time = perf_counter()
+
+    if not docker_client:
+        docker_client = docker.from_env()
+
+    # create directories for 'transformed' outputs
+    transformed_directory_a, transformed_directory_b = create_subdirectories(
+        base_directory=run_directory, subdirectories=["transformed/a", "transformed/b"]
+    )
+    logger.info(
+        "Transformed directories created: "
+        f"{[transformed_directory_a, transformed_directory_b]}"
+    )
+
+    # create list of run configs
+    run_configs = [
+        (image_name_a, transformed_directory_a),
+        (image_name_b, transformed_directory_b),
+    ]
+
+    # transform input files using images A and B via Docker containers
+    # save transformed record JSON files in corresponding directory
+    containers = []
+    for input_file in input_files:
+        for docker_image, transformed_directory in run_configs:
+            container = run_docker_container(
+                docker_image, transformed_directory, input_file, docker_client
+            )
+            containers.append(container)
+
+    # wait for all container runs to complete (status = "exited")
+    wait_for_containers(containers, timeout)
+    logger.info(f"All {len(containers)} containers have exited.")
+
+    # get logs from all Docker containers
```

Comment:
Hate to even ask, but what about breaking this out as a little standalone function as well, where you pass it the list of `containers`?

Reply:
See new function [`aggregate_logs`](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/6/files#diff-7a49d9dbf619fca67eb5ed5ed16b6c62440a4cc6898754cd835d2b819be74ddcR178-R191).